### PR TITLE
[RLlib] Make sure SlateQ works with GPU.

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

Create models and variables on proper device so SlateQ works with GPU training.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 1]
```diff
@@ -41,14 +41,15 @@ def build_slateq_model_and_distribution(
     Returns:
         Tuple consisting of 1) Q-model and 2) an action distribution class.
     """
+    device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
```
Comment (sven1977) [inline rllib/agents/slateq/slateq_torch_policy.py:44]: Actually, the TorchPolicy will take care of all this.

I think we only have to make sure all tensors in the loss function are on the right device. ...

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 2]
```diff
@@ -154,7 +155,7 @@ def build_slateq_losses(
 
     clicked = torch.sum(click_indicator, dim=1)
     mask_clicked_slates = clicked > 0
-    clicked_indices = torch.arange(batch_size)
+    clicked_indices = torch.arange(batch_size).to(policy.device)
```
Comment (sven1977) [inline rllib/agents/slateq/slateq_torch_policy.py:158]: Here, this seems (almost) correct.

```
clicked_indices = torch.arange(batch_size).to(clicked.device). # <- some tensor that we know is already on one of the GPU.
```

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 3]
```diff
@@ -320,7 +321,10 @@ def score_documents(
         torch.multiply(user_obs.unsqueeze(1), torch.stack(doc_obs, dim=1)), dim=2
     )
     # Compile a constant no-click score tensor.
-    score_no_click = torch.full(size=[user_obs.shape[0], 1], fill_value=no_click_score)
+    # Make sure it lives on the same device as scores_per_candidate.
+    score_no_click = torch.full(
```
Comment (sven1977) [inline rllib/agents/slateq/slateq_torch_policy.py:325]: looks great.

File: rllib/agents/slateq/slateq_torch_policy.py (hunk 1 above)
Comment (gjoliver) [reply rllib/agents/slateq/slateq_torch_policy.py:44]: ok, I understand everything now.
it's the target_model that was the issue. we just need to use the correct target_model out of policy.target_models for things to work.
thanks.

File: rllib/agents/slateq/slateq_torch_policy.py (hunk 2 above)
Comment (gjoliver) [reply rllib/agents/slateq/slateq_torch_policy.py:158]: yep. this was actually how I did it originally. :)
updated.

File: rllib/agents/slateq/slateq_torch_policy.py [hunk 4]
```diff
@@ -392,20 +394,34 @@ def postprocess_fn_add_next_actions_for_sarsa(
...
     Args:
         policy: The Policy object.
         obs_space: The Policy's observation space.
         action_space: The Policy's action space.
         config: The Policy's config.
     """
+    # [1, AxS] Useful for torch.take_along_dim()
+    policy.slates_indices = policy.slates.reshape(-1).unsqueeze(0).to(policy.device)
+
+    setup_mixins(policy)
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/agents/slateq/slateq_torch_policy.py:414]: ah, reverted. I was trying to move policy.slates_indices to the correct device during late_setup.
but I am doing this the correct way now.

File: rllib/agents/slateq/slateq_torch_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/slateq/slateq_torch_policy.py]: Ah, yeah, sorry, I should have thought about this. Yes, you can always do:
```
correct_target_model_to_use = policy.target_models[model]
...
```

**Review by sven1977 — APPROVED** [review]
Looks great! Thanks for this important fix @gjoliver !
